parlai/core/utils.py

```diff
@@ -713,6 +739,45 @@ def __contains__(self, key):
         """Determine if text contains any offensive words in the filter."""
         return self.contains_offensive_language(key)
 
+    def str_segment(self, text, dict_agent):
+
+        @self.memo
+        def segment(text):
+            # Return a list of words that is the best segmentation of text.
+            if not text:
+                return []
+            candidates = ([first]+segment(rem) for first, rem in splits(text))
+            return max(candidates, key=prob_words)
+
+        def splits(text, L=20):
+            # Returns a list of all possible first and remainder tuples where
+            # first +" "+ rem = tex
+            candidates = []
+            for i in range(min(len(text), L)):
+                candidates.append((text[:i+1], text[i+1:]))
+            return candidates
+
+        def prob_words(words):
+            # Returns Naive Bayes probability for a sequence of words
+            return product(prob_dist(w) for w in words)
```

Comment:
generally it's a bad idea to multiply probabilities, as you can easily underflow. You should instead store logprobs and add them.